python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -460,14 +467,68 @@ def get_latest_progress(self) -> "StepID":
         return asyncio_run(self._get(self._key_workflow_progress(),
                                      True))["step_id"]
 
+    def _put_object_ref(self, obj_ref: ObjectRef,
+                        upload_tasks: List[ObjectRef]) -> str:
+        @ray.remote
+        def put_helper(loc: str, obj: Any, wf_storage: WorkflowStorage) -> str:
+            asyncio.get_event_loop().run_until_complete(
+                wf_storage._put(loc, obj))
+
+        # TODO (Alex): Ideally we could parallelize these hash calculations,
+        # but we need the result before we can return from this reducer.
+        identifier = calculate_identifiers([obj_ref]).pop()
+        paths = self._key_obj_id(identifier)
+        loc = self._storage.make_key(*paths)
+        # TODO (Alex): We should dedupe these puts with the global coordinator.
+        task = put_helper.remote(loc, obj_ref, self)
+        upload_tasks.append(task)
+        return _load_object_ref, (loc, self)
+
     async def _put(self, paths: List[str], data: Any,
-                   is_json: bool = False) -> None:
+                   is_json: bool = False) -> str:
         try:
+            upload_tasks: List[ObjectRef] = []
+            if not is_json:
+                # Setup our custom serializer.
+                output_buffer = io.BytesIO()
+
+                # Cloudpickle doesn't support private dispatch tables, so we
+                # extend the cloudpickler instead to avoid changing
+                # cloudpickle's global dispatch table which is shared with
+                # `ray.put`. See
+                # https://github.com/cloudpipe/cloudpickle/issues/437
+                class ObjectRefPickler(cloudpickle.CloudPickler):
+                    _object_ref_reducer = {
+                        ray.ObjectRef: lambda ref: self._put_object_ref(
+                            ref, upload_tasks)
+                    }
+                    # _object_ref_reducer = {}
+                    dispatch_table = ChainMap(
+                        _object_ref_reducer,
+                        cloudpickle.CloudPickler.dispatch_table)
+                    dispatch = dispatch_table
+
+                pickler = ObjectRefPickler(output_buffer)
+                pickler.dump(data)
+                output_buffer.seek(0)
+                value = output_buffer.read()
+            else:
+                value = data
+
             key = self._storage.make_key(*paths)
-            await self._storage.put(key, data, is_json=is_json)
+
+            await self._storage.put(key, value, is_json=is_json)
```

Comment:
I think actually the dump is done inside self._storage. Is it possible to follow this pattern?

Reply:
For example, [here](https://github.com/ray-project/ray/blob/1685ce437aeab2bad89bc494504d5f85c23adaa0/python/ray/experimental/workflow/storage/filesystem.py#L136-L142) it's actually calling ray.cloudpickle.dump.

Reply:
The reason we put the reduce into the storage actually is because of memory concerns. Basically, storage will have the opportunity to avoid using a big chunks of data in memory.

For example, s3 storage, use file as temporary storage for big data, and fs storage dumps to files directly.

As discussed today, we probably will have customized put/get. Later we can just put a customized class. Can you confirm that I understand it correctly here? I'm OK with updating it in the following PR.

Reply:
I think I understand your concern about memory usage. I think the solution I'd want to do in the following PR is to add an API to the storage:

```
class Storage:
  async def put_file(self, key: str, data: typing.IO) -> None:
    ...
```

That way the storage doesn't handle serialization at all.